modules/costs/spec/requests/api/time_entries/create_resource_spec.rb

```diff
@@ -0,0 +1,124 @@
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+require "rack/test"
+
+RSpec.describe "API v3 Time Entries resource",
+               content_type: :json,
+               with_settings: { allow_tracking_start_and_end_times: true } do
+  include Rack::Test::Methods
+  include API::V3::Utilities::PathHelper
+
+  let(:project) { create(:project) }
+  let(:active_activity) { create(:time_entry_activity) }
+  let(:work_package) { create(:work_package, project:) }
+  let(:permissions) { %i[log_time view_work_packages] }
+
+  let(:german_user) do
+    create(:user, member_with_permissions: { project => permissions }, preferences: { time_zone: "Europe/Berlin" })
+  end
+  let(:japanese_user) do
+    create(:user, member_with_permissions: { project => [:view_project] }, preferences: { time_zone: "Asia/Tokyo" })
+  end
+
+  let(:path) { api_v3_paths.time_entries }
+  let(:parameters) { {} }
+  let(:json_response) { JSON.parse(response.body) }
+
+  subject(:response) { last_response }
+
+  describe "#POST /api/v3/time_entries" do
+    before do
+      login_as(german_user)
+    end
+
+    describe "correct time zone handling for start time" do
+      context "when logging time for yourself" do
+        let(:parameters) do
+          {
+            _links: {
+              workPackage: { href: api_v3_paths.work_package(work_package.id) },
+              project: { href: api_v3_paths.project(project.id) },
+              activity: { href: api_v3_paths.time_entries_activity(active_activity.id) }
+            },
+            spentOn: "2024-12-24",
+            hours: "PT2H",
+            startTime: "2024-12-24T12:00:00Z"
+          }
+        end
+
+        it "creates the time entry and sets the start time to the users timezone" do
+          post path, parameters.to_json
+
+          expect(subject).to have_http_status(:created)
+          time_entry_id = json_response["id"]
+          time_entry = TimeEntry.find(time_entry_id)
+
+          expect(time_entry.spent_on).to eq(Date.new(2024, 12, 24))
+          expect(time_entry.hours).to eq(2)
+          expect(time_entry.start_time).to eq(13 * 60) # 12:00 UTC = 13:00 Berlin on December 24th, 2024
+          expect(time_entry.time_zone).to eq("Europe/Berlin")
```

Comment:
❓ maybe stupid question, but how does the lib differentiate between DLS and standard time for `Europe/Berlin`? Is there another identifier for DLS time? Or might this test break in April?

Reply:
Europe/Berlin is a fixed identifier for a timezone that has rules attached to it with information when DST starts and when the offset to UTC changes. It then applies the correct offset when you pass it the date and time:

```
[1] pry(main)> tz = ActiveSupport::TimeZone['Europe/Berlin']
=> #<ActiveSupport::TimeZone:0x00000001628dbc70 @name="Europe/Berlin", @tzinfo=#<TZInfo::DataTimezone: Europe/Berlin>, @utc_offset=nil>

[2] pry(main)> tz.local(2024, 12, 24, 12, 0, 0)
=> Tue, 24 Dec 2024 12:00:00.000000000 CET +01:00

[3] pry(main)> tz.local(2024, 6, 18, 12, 0, 0)
=> Tue, 18 Jun 2024 12:00:00.000000000 CEST +02:00
```

As we use a fixed date here in the test (24.12.2024), we will always get a timestamp using Central European Time, never Central European Summer Time.

Reply:
That's also why it is always a good idea to use the local identifiers like Europe/Berlin, Asia/Seoul or America/Mexico_City instead of things like `CET`, `KST` or `CST`